drivers/ws281x/atmega.c

```diff
@@ -94,114 +98,119 @@ void ws281x_write_buffer(ws281x_t *dev, const void *buf, size_t size)
-            break;
+                break;
+            default:
+                assert(0);
+                break;
+        }
     }
-#elif (CLOCK_CORECLOCK >= 15500000U) && (CLOCK_CORECLOCK <= 16500000U)
-    while (pos < end) {
-        uint8_t cnt = 8;
-        uint8_t data = *pos;
-        while (cnt > 0) {
-            __asm__ volatile(                       /* CPU Cycles | 1 | 0 */
-                "st     %a[port], %[on]"    "\n\t"  /* 2          | x | x */
-                "sbrc   %[data], 7"         "\n\t"  /* 1          | x | x */
-                "rjmp   .+0"                "\n\t"  /* 2          | x | - */
-                "sbrc   %[data], 7"         "\n\t"  /* 1          | x | x */
-                "rjmp   .+0"                "\n\t"  /* 2          | x | - */
-                "sbrc   %[data], 7"         "\n\t"  /* 1          | x | x */
-                "nop"                       "\n\t"  /* 1          | x | - */
-                "st     %a[port], %[off]"   "\n\t"  /* 2          | x | x */
-                /* Total CPU cycles for high period:
-                 * 10 cycles for bit 1, 5 cycles for bit 0 */
-                : [data]    "+r" (data)
-                : [on]      "r" (mask_on),
-                  [port]    "e" (port_addr),
-                  [off]     "r" (mask_off)
-            );
-            cnt--;
-            data <<= 1;
+    else if (coreclk() >= 15500000U && coreclk() <= 16500000U) {
+        while (pos < end) {
+            uint8_t cnt = 8;
+            uint8_t data = *pos;
+            while (cnt > 0) {
+                __asm__ volatile(                       /* CPU Cycles | 1 | 0 */
+                    "st     %a[port], %[on]"    "\n\t"  /* 2          | x | x */
+                    "sbrc   %[data], 7"         "\n\t"  /* 1          | x | x */
+                    "rjmp   .+0"                "\n\t"  /* 2          | x | - */
+                    "sbrc   %[data], 7"         "\n\t"  /* 1          | x | x */
+                    "rjmp   .+0"                "\n\t"  /* 2          | x | - */
+                    "sbrc   %[data], 7"         "\n\t"  /* 1          | x | x */
+                    "nop"                       "\n\t"  /* 1          | x | - */
+                    "st     %a[port], %[off]"   "\n\t"  /* 2          | x | x */
+                    /* Total CPU cycles for high period:
+                     * 10 cycles for bit 1, 5 cycles for bit 0 */
+                    : [data]    "+r" (data)
+                    : [on]      "r" (mask_on),
+                      [port]    "e" (port_addr),
+                      [off]     "r" (mask_off)
+                );
+                cnt--;
+                data <<= 1;
+            }
+            pos++;
         }
-        pos++;
     }
-#else
-#error "No low level WS281x implementation for ATmega CPUs for your CPU clock"
-#endif
+    else {
+        DEBUG("No low level WS281x implementation for ATmega CPUs for your CPU clock\n");
+        assert(0);
+    }
```

Comment:
Please use a `static_assert()` to check this at compile-time. Emitting code even though the configuration is known to be broken at compile time and only print a warning about this at runtime with `ENABLE_DEBUG` set to 1 is IMO one of the worst anti-patterns used at many places in RIOT's code-base. IMO we really should try to stop introducing new instances of this debugging-time-sinkholes.

Otherwise I really like getting rid of a whole bunch of preprocessor statements here. (If the compiler frownes upon that `static_assert()`, you can use `DECLARE_CONSTANT()` as an impromptu `constexpr` alternative. With it, you create a constant of the boolean expression and can refer to that in the `static_assert()`.)

Reply:
using static_assert with the DECLARE_CONSTANT doesn't compile, not sure I know how to make it work. It seems the call to `coreclk()` function in the constexpr is not recognized as const.

Reply:
OK, I see. I'll try to find a way around.

Reply:
Both GCC and clang resist accessing the return value of a `static inline` function in a `static_assert()`. I only see two alternatives: Keep using `CLOCK_CORECLOCK` here (given that this driver is ATmega specific, this exception won't hurt that much) or abusing the linker:

```diff
diff --git a/drivers/ws281x/atmega.c b/drivers/ws281x/atmega.c
index 47d901e5fc..1181ca510b 100644
--- a/drivers/ws281x/atmega.c
+++ b/drivers/ws281x/atmega.c
@@ -23,12 +23,12 @@
 #include <stdint.h>
 #include <string.h>
 
-#include "clk.h"
 #include "atmega_gpio.h"
+#include "clk.h"
+#include "periph_cpu.h"
 #include "ws281x.h"
-#include "ws281x_params.h"
 #include "ws281x_constants.h"
-#include "periph_cpu.h"
+#include "ws281x_params.h"
 
 #define ENABLE_DEBUG 0
 #include "debug.h"
@@ -208,8 +208,11 @@ void ws281x_write_buffer(ws281x_t *dev, const void *buf, size_t size)
         }
     }
     else {
-        DEBUG("No low level WS281x implementation for ATmega CPUs for your CPU clock\n");
-        assert(0);
+        /* No low level WS281x implementation for ATmega CPUs for the used CPU
+         * clock. Abusing the linker to give a compile time error instead of
+         * emitting a broken firmware */
+        extern void clock_frequency_is_not_supported_by_ws281x_driver(void);
+        clock_frequency_is_not_supported_by_ws281x_driver();
     }
 }
 
```

This gives me for my (unsupported) 12 MHz out-of-tree board:

```
/usr/lib/gcc/avr/11.2.0/../../../../avr/bin/ld: /home/maribu/Repos/software/RIOT/tests/driver_ws281x/bin/m1284p/ws281x/atmega.o: in function `ws281x_write_buffer':
/home/maribu/Repos/software/RIOT/drivers/ws281x/atmega.c:215: undefined reference to `clock_frequency_is_not_supported_by_ws281x_driver'
collect2: error: ld returned 1 exit status
make: *** [/home/maribu/Repos/software/RIOT/tests/driver_ws281x/../../Makefile.include:686: /home/maribu/Repos/software/RIOT/tests/driver_ws281x/bin/m1284p/tests_driver_ws281x.elf] Error 1
```

But compiles just fine for the Arduino Mega2560. I'm not sure which hack is less ugly.

Reply:
I think that just keeping `assert(0);` and leaving a comment instead of `DEBUG` is enough. If it breaks at runtime, it's easy to know what happened with `DEBUG_ASSERT_VERBOSE` set afterwards.

Reply:
We had e.g. for the waspmote-pro an xtimer configuration that was broken for years, and all parameters needed to determine that it was broken were available at compile time. But since nobody added a compile time check, it went unnoticed for years. With a compile time check, Murdock would have caught it immediately.

IMO we should not compromise on compile-time checks. They are to helpful - especially since we mostly do compile checking only - to give up on them.

Reply:
Ok, let's used the ld script abuse (already squashed).